File: test/.work/robust_core/go/sb/struct_item.go
package sb

import (
	"bytes"
	"fmt"
	"math"
	"slices"
)

type Item struct {
	Id uint32  
	Color Color  
	Tags []string  
	Active bool  
}

func GetItem(buf *bytes.Buffer, s *Item) error {
	if s == nil { return nil }
	bitSize := int(math.Ceil(float64(4) / 8.0))
	if buf.Len() < bitSize { return fmt.Errorf("GetItem bitmask: %d - %d", buf.Len(), bitSize) }
	bits := buf.Next(bitSize)
	if GetBit(bits, uint8(0)) {
		val, err := GetU32(buf)
		if err != nil { return fmt.Errorf("GetItem Id: %w", err) }
		s.Id = val
	}
	if GetBit(bits, uint8(1)) {
		val, err := GetColor(buf)
		if err != nil { return fmt.Errorf("GetItem Color: %w", err) }
		s.Color = val
		if !IsColor(s.Color) { return fmt.Errorf("GetItem Color: 非法枚举值: %d", s.Color) }
	}
	if GetBit(bits, uint8(2)) {
		val, err := GetTextList(buf)
		if err != nil { return fmt.Errorf("GetItem Tags: %w", err) }
		s.Tags = val
	}
	s.Active = GetBit(bits, uint8(3))
	if err := ValidateItem(s); err != nil { return fmt.Errorf("ValidateItem: %w", err) }
	return nil
}

func SetItem(buf *bytes.Buffer, s *Item) error {
	if s == nil { return fmt.Errorf("SetItem: nil value") }
	if err := ValidateItem(s); err != nil { return fmt.Errorf("ValidateItem: %w", err) }
	bits := make([]byte, uint8(math.Ceil(float64(4)/8.0)))
	body := bytes.NewBuffer(nil)
	if s.Id != 0 {
		if err := SetU32(body, s.Id); err != nil { return fmt.Errorf("SetItem Id: %w", err) }
		SetBit(bits, uint8(0), true)
	}
	if s.Color != 0 {
		if err := SetColor(body, s.Color); err != nil { return fmt.Errorf("SetItem Color: %w", err) }
		SetBit(bits, uint8(1), true)
	}
	if len(s.Tags) > 0 {
		if err := SetTextList(body, s.Tags); err != nil { return fmt.Errorf("SetItem Tags: %w", err) }
		SetBit(bits, uint8(2), true)
	}
	SetBit(bits, uint8(3), s.Active)

	if _, err := buf.Write(bits); err != nil { return fmt.Errorf("SetItem write bitmask: %w", err) }
	_, err := body.WriteTo(buf); return err
}

func ValidateItem(s *Item) error {
	if s == nil { return nil }
	if s.Color != 0 && !IsColor(s.Color) { return fmt.Errorf("Color 非法枚举值: %d", s.Color) }
	return nil
}

func EqItem(a, b *Item) bool {
	if a == b { return true }
	if a == nil || b == nil { return false }
	if !EqU32(a.Id, b.Id) { return false }
	if a.Color != b.Color { return false }
	if !EqTextList(a.Tags, b.Tags) { return false }
	if !EqBool(a.Active, b.Active) { return false }
	return true
}

// Standalone functions
func ReadItem(buf *bytes.Buffer) (*Item, error) {
	s := new(Item)
	return s, GetItem(buf, s)
}

type ItemList []*Item
func GetItemList(buf *bytes.Buffer) (ItemList, error) { return getList[*Item, ItemList](buf, ReadItem) }
func SetItemList(buf *bytes.Buffer, v ItemList) error { return setList(buf, v, SetItem) }
func ValidateItemList(v ItemList) error {
	for i, item := range v {
		if item == nil { return fmt.Errorf("ItemList[%d]: nil item", i) }
		if err := ValidateItem(item); err != nil { return fmt.Errorf("ItemList[%d]: %w", i, err) }
	}
	return nil
}
func EqItemList(a, b ItemList) bool { return slices.EqualFunc(a, b, EqItem) }

File: test/.work/robust_core/ts/sb/struct_item.ts
import * as _ from "./_"
import * as Enum from "./enum"

export interface Item extends _.Serializable, _.Deserializable {
    id: number;
    color: Enum.Color;
    tags: string[];
    active: boolean;
}

export const newItem = (): Item => {
    const s = {
        id: 0,
        color: 0,
        tags: [],
        active: false,
    } as any as Item;
    s.set = (buf: _.Buffer) => setItem(buf, s);
    s.get = (buf: _.Buffer) => {
        const [res, err] = getItem(buf);
        if (err === null) Object.assign(s, res);
        return err;
    };
    return s;
}

export const eqItem = (a: Item, b: Item): boolean => {
    if (a === b) return true;
    if (a === null || b === null) return false;
    if (!_.eqU32(a.id, b.id)) return false;
    if (a.color !== b.color) return false;
    if (!_.eqTextList(a.tags, b.tags)) return false;
    if (!_.eqBool(a.active, b.active)) return false;
    return true;
}

export const getItem = (buf: _.Buffer): [Item, Error | null] => {
    const s = newItem();
    const bitmaskSize = Math.ceil(4 / 8);
    const [bits, err] = buf.read(bitmaskSize);
    if (err !== null) return [s, err];
    if (_.GetBit(bits, 0)) {
        const [v, err] = _.getU32(buf);
        if (err !== null) return [s, err];
        s.id = v;
    }
    if (_.GetBit(bits, 1)) {
        const [v, err] = _.getU8(buf);
        if (err !== null) return [s, err];
        s.color = v as any;
        if (!_.IsColor(s.color as any)) return [s, new Error("get Item Color: invalid enum value")];
    }
    if (_.GetBit(bits, 2)) {
        const [v, err] = _.getTextList(buf);
        if (err !== null) return [s, err];
        s.tags = v;
    }
    s.active = _.GetBit(bits, 3);
    return [s, null];
}

export const setItem = (buf: _.Buffer, s: Item): Error | null => {
    if (s === null || s === undefined) return new Error(`set Item: value is null or undefined`);
    const bits = new Uint8Array(Math.ceil(4 / 8));
    const body = new _.Buffer();
    if (!_.eqU32(s.id, 0)) {
        const err = _.setU32(body, s.id);
        if (err !== null) return err;
        _.SetBit(bits, 0, true);
    }
    if ((s.color as any) !== 0) {
        if (!_.IsColor(s.color as any)) return new Error("set Item Color: invalid enum value");
        const err = _.setU8(body, s.color as any);
        if (err !== null) return err;
        _.SetBit(bits, 1, true);
    }
    if (s.tags && s.tags.length > 0) {
        const err = _.setTextList(body, s.tags);
        if (err !== null) return err;
        _.SetBit(bits, 2, true);
    }
    _.SetBit(bits, 3, s.active as boolean);

    const errBits = buf.write(bits);
    if (errBits !== null) return errBits;
    return buf.write(body.bytes);
}

export const getItemList = (buf: _.Buffer): [Item[], Error | null] => _.getList(buf, getItem);
export const setItemList = (buf: _.Buffer, v: Item[]): Error | null => _.setList(buf, v, setItem);
export const eqItemList = (a: Item[], b: Item[]): boolean => _.eqList(a, b, eqItem);
